### ISSUE TITLE
Property allownone is confusing

### ISSUE THREAD

**Issue by dessant** [kivy/kivy#3743]
Related chat: https://botbot.me/freenode/kivy-dev/2015-11-02/?msg=53292381&page=1
The default value of ObjectProperty is None, despite ObjectProperty(allownone=False).

We should discuss here the reason for having this option, and how could it be made more consistent, considering different use cases.

Some of us consider that setting allownone to True by default would be an improvement, at least for ObjectProperty.

**Comment by kived** [comment]
I don't see any reason to remove it, it does have uses, just like the `NOT NULL` column option in MySQL (and other RDBMS) or `Nullable<?>` types in C# (and similar constructions in Java). I mostly prefer the default `allownone=False` because it guarantees that the property contains a value with a specific type (default property values, which are not checked, notwithstanding). But I do agree that it could make more sense to default it to True for ObjectProperty only, as None is more likely to be used with an ObjectProperty and we don't do any default property checking (which would make `ObjectProperty(allownone=False)` raise an exception for the default None value).

The primary reason for doing this checking at assignment time, rather than letting any value be assigned and being type-agnostic like Python is in general, is that any resulting error will often occur separately from the assignment (i.e. if it is used in kv). All you know is that a certain property has been set to None, and you can see the error from trying to use that None, but with no clear indication of where the None actually came from.

I would otherwise be open to an argument to change the default to True for all property types, but the `allownone=False` default is well established at this point.

**Comment by matham** [comment]
I agree that having allownone and then allowing None is good for place where you want to indicate that it has no value, like `NULL`. In that context it makes sense, but I do think it should default to `True` for everything because otherwise it'll become confusing treating ObjectProperty as special. 

Also, I don't remember ever relying on it defaulting to `False`, but I do use `allownone=True` in quite a few places. Even for e.g. numeric properties etc.
